feat: add custom private dns zone name in networkSpec (#1289)
kubernetes-sigs/cluster-api-provider-azure: docs/book/src/topics/custom-dns.md

# Custom Private DNS Zone Name

> It is possible to set the Private DNS Zone Name in the NetworkSpec to change the Private DNS zone name by default the dns zone name is `${CLUSTER_NAME}.capz.io`

Reviewer: ```suggestion
It is possible to set the Private DNS Zone Name to a custom value by setting `PrivateDNSZoneName` in the `NetworkSpec`. By default the dns zone name is `${CLUSTER_NAME}.capz.io`
```

> It is possible to set the Private DNS Zone Name to a custom value by setting `PrivateDNSZoneName` in the `NetworkSpec`. By default the dns zone name is `${CLUSTER_NAME}.capz.io`

Reviewer 2: ```suggestion
It is possible to set the DNS zone name to a custom value by setting `PrivateDNSZoneName` in the `NetworkSpec`. By default the DNS zone name is set to `${CLUSTER_NAME}.capz.io`.
```

> *This feature is enable only if the `apiServerLB.type` is `Internal`*

Reviewer: ```suggestion
*This feature is enabled only if the `apiServerLB.type` is `Internal`*
```

Author: I apply all of them, thanks for the review :smile:

> *This feature is enabled only if the `apiServerLB.type` is `Internal`*

Reviewer 3: Should we also enforce this in validateCluster()?

Reviewer: +1 for validating in webhooks (better user feedback) even though it probably won't cause any undue side effects because we also check for `s.IsAPIServerPrivate()` before setting this.

Reviewer: @cyril-corbon can you plug a validation in this [method](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/7b8e79295d82d0371bbd82844ff189bb79e3f41f/api/v1alpha4/azurecluster_validation.go#L44) to make sure `PrivateDNSZoneName` is only set when api lb type is `Internal`?

Author: @CecileRobertMichon  @shysank I added a function to validate the Name and the lbType + some unit test  :smile: 
To check the name I use the following function : https://pkg.go.dev/github.com/asaskevich/govalidator?utm_source=godoc#IsDNSName
Can we import this lib or should we set a regex in const ? :thinking:

Reviewer 3: Sounds great, and thanks for adding tests!

> Can we import this lib or should we set a regex as const ? 🤔

yes, as long as it meets the requirements in https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md (from a quick scan it looks like it does)
